### ISSUE TITLE
Type guard from if statement doesn't apply to lambda statements

### ISSUE THREAD

**Issue by AaronSartin** [python/mypy#18099]
**Bug Report**

Type guards (at least from if statements) don't apply to lambda statements

**To Reproduce**

```python
from typing import AbstractSet, Callable, Optional


def get_filter_function(allowed_values: Optional[AbstractSet[int]]) -> Callable[[int, bool]:
    if allowed_values is None:
        return lambda _: True
    else:
        _ = 5 in allowed_values  # does not raise
        return lambda value: value in allowed_values  # raises Unsupported right operand type for in ("Optional[AbstractSet[int]]")
```

**Expected Behavior**

Neither line should be a mypy error.

**Actual Behavior**

The lambda raises an operator error ``Unsupported right operand type for in ("Optional[AbstractSet[int]]")``.

**Your Environment**

Originally observed on:

- Mypy version used: 1.11.2
- Python version used: 3.8.19

~~Replicated on:~~ Not able to replicate in:

- Mypy version used: 1.13.0
- Python version used: 3.13.0
